### PR TITLE
PRMDR-688 Virus scanner lambda

### PR DESCRIPTION
**Pull request by NogaNHS**

### CODE REVIEW

File: lambdas/services/virus_scan_result_service.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    def get_new_access_token(self):
+        try:
+            logger.info(f"username: {self.username}")
+            logger.info(f"password: {self.password}")
+
+            json_login = json.dumps(
+                {"username": self.username, "password": self.password}
+            )
+            token_url = self.base_url + "/api/Token"
```
Comment (SRAlexander) [inline lambdas/services/virus_scan_result_service.py:89]: Take the "/api/Token" section and replace it with a fixed variable

File: lambdas/services/virus_scan_result_service.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+            response = requests.post(
+                url=token_url,
+                headers={"Content-type": "application/json"},
+                data=json_login,
+            )
+
+            response.raise_for_status()
+            logger.info(response.json())
+            new_access_token = response.json()["accessToken"]
```
Comment (SRAlexander) [inline lambdas/services/virus_scan_result_service.py:99]: Do we need any protections in the event accessToken is not returned?

File: lambdas/services/virus_scan_result_service.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        self.staging_s3_bucket_name = os.getenv("STAGING_STORE_BUCKET_NAME")
+        self.ssm_service = SSMService()
+        self.username = ""
+        self.password = ""
+        self.base_url = ""
+        self.access_token = ""
+
+    def prepare_request(self, file_ref):
+        try:
+            self.get_ssm_parameters_for_request_access_token()
```
Comment (SRAlexander) [inline lambdas/services/virus_scan_result_service.py:29]: Would this fit better in the service __init__?

File: lambdas/services/virus_scan_result_service.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            self.virus_scan_request(file_ref, retry_on_expired=True)
```
Comment (SRAlexander) [inline lambdas/services/virus_scan_result_service.py:30]: rename to request_virus_scan

File: lambdas/services/virus_scan_result_service.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            )
+            raise VirusScanResultException(500, LambdaError.VirusScanAWSFailure)
+
+    def virus_scan_request(self, file_ref: str, retry_on_expired: bool):
+        try:
+            headers = {
+                "Content-Type": "application/json",
+                "Authorization": "Bearer " + self.access_token,
+            }
+            scan_url = self.base_url + "/api/Scan/Existing"
```
Comment (SRAlexander) [inline lambdas/services/virus_scan_result_service.py:44]: extract the endpoint string into a constant variable

File: lambdas/services/virus_scan_result_service.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                "objectPath": file_ref,
+            }
+            logger.info(json_data_request)
+
+            response = requests.post(
+                url=scan_url, data=json.dumps(json_data_request), headers=headers
+            )
+            if response.status_code == 401 and retry_on_expired:
+                self.get_new_access_token()
+                return self.virus_scan_request(file_ref, retry_on_expired=False)
```
Comment (SRAlexander) [inline lambdas/services/virus_scan_result_service.py:56]: I love a bit of recursion, nicely identified :)

File: lambdas/handlers/virus_scan_result_handler.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    try:
+        event_body = json.loads(event["body"])
+        if not event_body:
+            raise VirusScanResultException(400, LambdaError.VirusScanNoBody)
+
+        document_reference = event_body["documentReference"]
+    except (JSONDecodeError, KeyError):
+        raise VirusScanResultException(400, LambdaError.VirusScanNoBody)
+
+    virus_scan_result_service = VirusScanResultService()
```
Comment (SRAlexander) [inline lambdas/handlers/virus_scan_result_handler.py:41]: Could remove "result" from the variable name.

**Review by SRAlexander — CHANGES_REQUESTED** [review]

File: lambdas/services/virus_scan_result_service.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    def get_ssm_parameters_for_request_access_token(self):
+        access_token_key = SSMParameter.VIRUS_API_ACCESS_TOKEN.value
+        username_key = SSMParameter.VIRUS_API_USER.value
+        password_key = SSMParameter.VIRUS_API_PASSWORD.value
+        url_key = SSMParameter.VIRUS_API_BASE_URL.value
+
+        parameters = [username_key, password_key, url_key, access_token_key]
+
+        ssm_response = self.ssm_service.get_ssm_parameters(
```
Comment (SRAlexander) [inline lambdas/services/virus_scan_result_service.py:127]: Do we need to safeguard for an error here?

File: lambdas/services/virus_scan_result_service.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            parameter_type="SecureString",
+        )
+
+    def get_ssm_parameters_for_request_access_token(self):
+        access_token_key = SSMParameter.VIRUS_API_ACCESS_TOKEN.value
+        username_key = SSMParameter.VIRUS_API_USER.value
+        password_key = SSMParameter.VIRUS_API_PASSWORD.value
+        url_key = SSMParameter.VIRUS_API_BASE_URL.value
+
+        parameters = [username_key, password_key, url_key, access_token_key]
```
[the remark this answers is not in the document]
Comment (NogaNHS) [reply lambdas/services/virus_scan_result_service.py:125]: I'm not sure I understand the need. Why to call this function if not to reassign those variables?

File: lambdas/services/virus_scan_result_service.py (hunk 3 above)
Comment (NogaNHS) [reply lambdas/services/virus_scan_result_service.py:29]: I don't think so. As we wrap it with a try block here, and catch ClientError.

File: lambdas/services/virus_scan_result_service.py (hunk 2 above)
Comment (NogaNHS) [reply lambdas/services/virus_scan_result_service.py:99]: added a key error to the except

**Review by SRAlexander — APPROVED** [review]

**Pull request merged**